src/main/java/org/sonarlint/intellij/analysis/JavaAnalysisConfigurator.java

```diff
@@ -59,92 +68,158 @@
-      if (path != null) {
-        properties.put(JAVA_BINARIES_PROPERTY, path);
-        testPath = (testPath != null) ? (testPath + SEPARATOR + path) : path;
+    processCompilerOutput(moduleClasspath, module, topLevel, testClasspathOnly);
+    final ModuleRootManager mrm = ModuleRootManager.getInstance(module);
+    final OrderEntry[] orderEntries = mrm.getOrderEntries();
+    for (final OrderEntry entry : orderEntries) {
+      if (!topLevel && !isExported(entry)) {
+        continue;
+      }
+      if (entry instanceof ModuleOrderEntry) {
+        processModuleOrderEntry(moduleClasspath, testClasspathOnly, (ModuleOrderEntry) entry);
+      } else if (entry instanceof LibraryOrderEntry) {
+        processLibraryOrderEntry(moduleClasspath, (LibraryOrderEntry) entry, testClasspathOnly);
+      } else if (entry instanceof JdkOrderEntry) {
+        processJdkOrderEntry(moduleClasspath, ((JdkOrderEntry) entry).getJdk());
       }
     }
+  }
 
-    if (testPath != null) {
-      properties.put(JAVA_TEST_BINARIES_PROPERTY, testPath);
+  private static void processModuleOrderEntry(JavaModuleClasspath moduleClasspath, boolean testClasspathOnly, ModuleOrderEntry moduleOrderEntry) {
+    Module dependentModule = moduleOrderEntry.getModule();
+    Set<Module> dependentModules = (testClasspathOnly || isOnlyForTestClasspath(moduleOrderEntry.getScope())) ? moduleClasspath.testDependentModules()
+      : moduleClasspath.dependentModules();
+    if (!dependentModules.contains(dependentModule)) {
+      // Protect against circular dependencies
+      dependentModules.add(dependentModule);
+      collectModuleClasspath(moduleClasspath, dependentModule, false, testClasspathOnly || isOnlyForTestClasspath(moduleOrderEntry.getScope()));
     }
   }
 
-  private static void configureLibraries(Module ijModule, Map<String, String> properties) {
-    List<String> libs = new ArrayList<>();
-    for (VirtualFile f : getProjectClasspath(ijModule)) {
-      libs.add(toFile(f.getPath()));
+  private static void processLibraryOrderEntry(JavaModuleClasspath moduleClasspath, LibraryOrderEntry libraryOrderEntry, boolean testClasspathOnly) {
+    Library lib = libraryOrderEntry.getLibrary();
+    final boolean libOnlyForTestClasspath = testClasspathOnly || isOnlyForTestClasspath(libraryOrderEntry.getScope());
+    getLibraryEntries(lib)
+      .stream()
+      .map(VfsUtilCore::virtualToIoFile)
+      .map(File::getAbsolutePath)
+      .forEach(libPath -> {
+        if (libOnlyForTestClasspath) {
+          moduleClasspath.testLibraries().add(libPath);
+        } else {
+          addProductionClasspathEntry(moduleClasspath, libPath);
+        }
+      });
+  }
+
+  private static boolean isOnlyForTestClasspath(DependencyScope scope) {
+    return !scope.isForProductionRuntime() && !scope.isForProductionCompile();
+  }
+
+  private static void processCompilerOutput(JavaModuleClasspath moduleClasspath, @NotNull Module module, boolean topLevel, boolean testModule) {
```

Comment:
I think these flags could also be reworked, it would simplify the nesting of conditionals